### PR TITLE
[Bugfix] Repeat peeks until quiescence

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Title says all

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3792

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by lukasz-golebiewski** [comment]
bors try

**Review by dzajkowski — APPROVED** [review]

**Comment by lukasz-golebiewski** [comment]
bors retry

File: rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala [hunk 1]
```diff
@@ -243,18 +243,20 @@ trait ReplayRSpaceTests extends ReplayRSpaceTestsBase[String, Pattern, String, S
                            false,
                            peeks = SortedSet(0)
                          )
-        resultProduce1 <- space.produce(channels(1), datum, false)
-        resultProduce2 <- space.produce(channels(0), datum, false)
+        resultProduce1  <- space.produce(channels(1), datum, false)
+        resultProduce2  <- space.produce(channels(0), datum, false)
+        resultProduce2a <- space.produce(channels(0), datum, false)
```
Comment (ArturGajowy) [inline rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala:248]: This value is unused, just like the other newly introduced in this test.
We should probably add assertions for the newly-captured values.

In general, I'd say this test is quite big and not easy to understand. Maybe having helper methods that assert the result of the produce/consume while they perform the produce/consume would help. I share this for the long term, all suggestions in this paragraph are optional for this PR :)

File: rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala [hunk 2]
```diff
@@ -275,18 +277,20 @@ trait ReplayRSpaceTests extends ReplayRSpaceTestsBase[String, Pattern, String, S
                            false,
                            peeks = SortedSet(0)
                          )
-        replayProduce1 <- replaySpace.produce(channels(1), datum, false)
-        replayProduce2 <- replaySpace.produce(channels(0), datum, false)
+        replayProduce1  <- replaySpace.produce(channels(1), datum, false)
+        replayProduce2  <- replaySpace.produce(channels(0), datum, false)
+        resultProduce2a <- replaySpace.produce(channels(0), datum, false)
```
Comment (ArturGajowy) [inline rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala]: wow, I wasn't even aware that you can do that in a single `for`:

```scala
for {
x <- Task.unit
x <- Task.unit
} yield ()
```

No surprise given the desugaring, but I'd expect some linter to yell at this :joy:

File: rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala [hunk 3]
```diff
@@ -471,24 +483,18 @@ trait ReplayRSpaceTests extends ReplayRSpaceTestsBase[String, Pattern, String, S
         val produces = Random.shuffle(channels)
         def consumeAndProduce(s: ISpace[Task, String, Pattern, String, String]) =
           for {
-            r  <- s.consume(channels, patterns, continuation, false, peeks = peeks)
-            rs <- produces.traverse(ch => s.produce(ch, s"datum-$ch", false))
+            r <- s.consume(channels, patterns, continuation, false, peeks = peeks)
+            rs <- produces.traverse { ch =>
+                   s.produce(ch, s"datum-$ch", false)
+
+                 }
```
Comment (ArturGajowy) [inline rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala]: please restore the previous formatting :)

File: rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala [hunk 4]
```diff
@@ -471,24 +483,18 @@ trait ReplayRSpaceTests extends ReplayRSpaceTestsBase[String, Pattern, String, S
...
           rs         <- consumeAndProduce(space)
           _          = rs.flatten should have size 1
           hs         = store.get()
-          _ <- channelsRange.traverse { i =>
-                val ch = s"channel$i"
-                hs.getData(ch).map { data =>
-                  if (peeks.contains(i))
-                    data should have size 1
-                  else
-                    data should have size 0
```
Comment (ArturGajowy) [inline rspace/src/test/scala/coop/rchain/rspace/ReplayRSpaceTests.scala:489]: shouldn't we retain `data should have size 0` assertion for all channels?

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala [hunk 5]
```diff
@@ -213,6 +213,35 @@ trait StorageActionsTests[F[_]]
       _             = insertActions should have size 0
     } yield ()
   }
+  "consuming and then producing on the same channel with persistent flag" should
+    "return the continuation and data and remove the persistent data" in fixture {
```
Comment (ArturGajowy) [inline rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: Shouldn't this say

>  "return the continuation and data and **not insert** the persistent data"

This gave me chills :joy:

File: rholang/src/test/scala/coop/rchain/rholang/PeekSpec.scala [hunk 6]
```diff
@@ -77,4 +77,37 @@ class PeekSpec extends FlatSpec with Matchers {
...
   }
+
+  it should "clear all peeks when inserting a persistent send" in {
+    mkRuntime[Task](tmpPrefix)
+      .use { runtime =>
+        implicit val c = runtime.cost
+        for {
+          _   <- evaluate[Task](runtime, """for (_ <<- @0) { @1!(0) }""")
+          _   <- evaluate[Task](runtime, """for (_ <<- @0) { @1!(0) }""")
+          _   <- evaluate[Task](runtime, """@0!!(0)""")
```
Comment (ArturGajowy) [inline rholang/src/test/scala/coop/rchain/rholang/PeekSpec.scala:88]: May I ask a test like this one, but with `@0!(0)` too? :)
("clear all peeks when inserting a(ny) send")

**Review by ArturGajowy — CHANGES_REQUESTED** [review]
Couple minors, GJ overall. Thanks for handling this!

File: rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala (hunk 5 above)
Comment (lukasz-golebiewski) [reply rspace/src/test/scala/coop/rchain/rspace/StorageActionsTests.scala]: You're right!

File: rholang/src/test/scala/coop/rchain/rholang/PeekSpec.scala (hunk 6 above)
Comment (lukasz-golebiewski) [reply rholang/src/test/scala/coop/rchain/rholang/PeekSpec.scala:88]: Can do

**Review by ArturGajowy — APPROVED** [review]

**Pull request merged**